src/solvers/smt2_incremental/construct_value_expr_from_smt.cpp

```diff
@@ -39,7 +39,7 @@ class value_expr_from_smt_factoryt : public smt_term_const_downcast_visitort
   {
     if(
       const auto integer_type =
```

Comment:
Should the variable name be changed then, if I no longer strictly corresponds to an integer type?

Reply:
Wait I'm reading more on this, and I'm becoming more and more confused...

It seems that `bool` maps to `_Bool`, which is a integral type with a two value domain: `{0, 1}`. If that's the case, why is it not transferable to an `integer`?

Is it because we have messed up the handling for that internally (in CBMC that is)?

Reply:
To answer your original question - Yes. I should rename this.

`_Bool` is a type which can only be set to 0 or 1 - "two value domain: {0, 1}". However you can expect that it will be stored in 1 or more bytes of memory, rather than packed into individual bits. The (unused) most significant bits will then be packed with `0`s to fill the full piece of memory allocated. For example if `_Bool` is 8 bits wide then the values will be `00000001` and `00000000` for true and false respectively. The property of allowing values 0 and 1 exclusively is satisfied by mapping all non-zero values to 1 when casting. So `(bool)0xff` will evaluate to `00000001` in binary.

I have put together a short test program which applies arithmetic operations to boolean values and the result is the same for concrete evaluation using `gcc` to compile and run and with running the same program though `cbmc`. Therefore the behaviour appears to be correct from the perspective of a user of cbmc. Whether this is (A) undefined behaviour according to the C standard. (B) Implicitly casting booleans to integer types to perform the arithmetic or (C) applying the arithmetic operations directly to the boolean values, isn't a question I have an answer to yet.

In conclusion - whether `c_bool_typet` should derive from `bitvector_typet` as it does at the moment, or from `integer_bitvector_typet` or some other type is an open question to me. For what it is worth, I'd like to limit the scope of this PR to supporting the current data structures rather than a wider refactor of CBMC.

Reply:
Agree with you re the scope of this.

Your answer satisfied my question.